zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -137,15 +158,74 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
         );
 
         // Transfer value from caller to callee
+        let intrinsic_tx_value = cb.query_word();
+        let intrinsic_mul_gas_fee_by_gas = cb.query_word();
+        let intrinsic_tx_value_and_gas_fee_is_zero = BatchedIsZeroGadget::construct(
+            cb,
+            [
+                intrinsic_tx_value.expr(),
+                intrinsic_mul_gas_fee_by_gas.expr(),
+            ],
+        );
+        cb.condition(is_tx_invalid.expr(), |cb| {
+            cb.require_equal(
+                "intrinsic_tx_value == 0 && intrinsic_mul_gas_fee_by_gas == 0",
+                intrinsic_tx_value_and_gas_fee_is_zero.expr(),
+                true.expr(),
+            );
+        });
+
+        cb.condition(1.expr() - is_tx_invalid.expr(), |cb| {
+            cb.require_equal(
+                "intrinsic_tx_value == tx_value",
+                intrinsic_tx_value.expr(),
+                tx_value.expr(),
+            );
+            cb.require_equal(
+                "intrinsic_tx_value == tx_value",
+                intrinsic_mul_gas_fee_by_gas.expr(),
+                mul_gas_fee_by_gas.product().expr(),
+            );
+        });
+
         let transfer_with_gas_fee = TransferWithGasFeeGadget::construct(
             cb,
             tx_caller_address.expr(),
             tx_callee_address.expr(),
-            tx_value.clone(),
-            mul_gas_fee_by_gas.product().clone(),
+            intrinsic_tx_value.clone(),
+            intrinsic_mul_gas_fee_by_gas.clone(),
             &mut reversion_info,
         );
 
+        // Verify transfer
+        let sender_balance_prev = transfer_with_gas_fee.sender.balance_prev();
+
+        let add_tx_value_and_mul_gas_fee_by_gas = cb.query_word();
+
+        // tx_value.int_value + gas_fee.int_value
+        let add_tx_value_and_mul_gas_fee_by_gas_gadget = AddWordsGadget::construct(
+            cb,
+            [tx_value.clone(), mul_gas_fee_by_gas.product().clone()],
+            add_tx_value_and_mul_gas_fee_by_gas.clone(),
+        );
+
+        // balance_not_enough = sender_balance_prev.int_value < tx_value.int_value +
+        // gas_fee.int_value
+        let balance_not_enough = LtWordGadget::construct(
+            cb,
+            &sender_balance_prev,
+            &add_tx_value_and_mul_gas_fee_by_gas,
+        );
+
+        // invalid_tx = 1 - (1 - balance_not_enough) * (is_nonce_valid)
+        // prover should not give incorrect is_tx_invalid flag.
+        // instruction.constrain_equal(is_tx_invalid, invalid_tx)
+        cb.require_equal(
+            "is_tx_invalid is correct",
+            1.expr() - (1.expr() - balance_not_enough.expr()) * is_nonce_valid.expr(),
```

Comment:
```suggestion
            or::expr(balance_not_enough.expr(), not::expr(is_nonce_valid.expr())),
```

Reply:
Updated. It is simpler. I didn't know `or::expr` and `not::expr`. Thanks!